### DISCUSSION TITLE
Strategies for off-centered object rotation

### DISCUSSION THREAD

**Discussion by theredsox** [fabricjs/fabric.js#9968] [General]
Peeking at the source code for rotate(), I turned that into an animation. Then wanted to tweak it to support off-centered rotation. I came up with [this JSFiddle](https://jsfiddle.net/dugafshq/1/). It takes a rectangle, an offset, and for every click of the button rotates the rectangle 90 degrees around the initial offset point. I'm curious:

- Is there a better way to do this than calculating the center point of the rectangle for each angle the animation triggers on? 
- Is there an offset property I could set, like rotatingPointOffset, but for programmatic rotation? 
- Can [transformations](http://fabricjs.com/using-transformations) handle it? Looking at the boss/minion tutorial, they rotate around the boss without doing raw calculations. Though I don't understand that magic yet to adapt it for my case.
- [An old SO from 2018](https://stackoverflow.com/questions/48290646/fabric-js-custom-rotation-point) describes using object.centeredRotation = false and transformation stuff. In the latest version, would it be something like call object.calcTransformMatrix(), change the angle in the matrix, and then call something to pull out a new center point to set in the animation onchange?

What I have looks like it will work, so not a huge deal, but guessing there is a better way and would like to learn. Any recommendations are appreciated.

**Comment by theredsox** [comment]
Ok, so actually implementing the Fiddle solution in my app shows that it doesn't fully work. Hitting the same issue as other less elegant approaches I've tried. Certain pivot offsets don't adjust properly for the whole rotation. See [Fiddle for pivot offset (-50, -25)](https://jsfiddle.net/dugafshq/2/). Often when they don't work fully, 2 of the 4 rotation positions seem off. Is it a geometry thing I'm screwing up? 

I feel like the issue lies in lines 44-56. Too simplistic for every combo of center position, pivot position, and angle. Because top and left change perspectives as the object rotates, it is confusing for me to determine what change is needed in that area. Any ideas?

**Reply by theredsox** [reply]
Alrighty, figured out the Fiddle issue. It wasn't lines 44-56. When the animation onchange does the cos/sin calc from starting angle to ending angle, calculating the current center, I needed to take into account the initial angle between the center point and the pivot point. That's why when I changed the pivot to have a different Y coord it would stop working. Because the initial angle between the two no longer happened to be 0. The onchange calculations need to offset by the initial angle. So even if the rectangle is starting at 180 and going to 270, I need to know the offset angle when its at 0 to use in the calcs.

[See new Fiddle](https://jsfiddle.net/dugafshq/3/) for the modified solution. The key is the new determineOffsetAngle() function triggered before an animation starts that takes the inverse tangent of (cY - pY)/(cX - pX) where c is center point of rectangle and p is the pivot point the animation is going to use. That resulting angle is added to every calculation done in the onChange as the animation goes between angle A to B... because really we're going between A + offset relative to pivot to B + offset relative to pivot.

I'd still love to know how to shortcut this work with other fabric lib techniques if such exist. I imagine they are more readable.

**Reply by theredsox** [reply]
Continuing with the fun, the Fiddle was still not working in certain cases like offset (50, 25) 😢. Two more changes 1) Wrong equation in determineOffsetAngle(), needed atan2() to get an angle between 0 and 360. 2) I've been confusing the center of the fabric object with the center of the arc. So was feeding the points into the equations in the wrong order.

Here's a set of working Fiddle tests covering a point in each quadrant:
[Offset (50, 25)](https://jsfiddle.net/dugafshq/4/)
[Offset (-50, -25)](https://jsfiddle.net/dugafshq/5/)
[Offset (50, -25)](https://jsfiddle.net/dugafshq/6/)
[Offset (-50, 25)](https://jsfiddle.net/dugafshq/7/)

So hopefully this at least functions when fed into my game 🤞.

**Reply by asturur** [reply]
Hi @theredsox rotation is a mix of changing angle and changing the position of the object.
originX and originY will go away at some point ( next major version read here: https://github.com/fabricjs/fabric.js/discussions/9736 )
https://jsfiddle.net/asturur/o1w5uczn/7/
This fiddle should be easily translatable to your current fabricjs version and is very simple compared to yours and does not use target.originX and target.originY.

Regarding your question if transformations can hanlde this, i do not think so, the transformation code is so much dependent by continuous events like mouse, triggering them programmatically is probably unnecessary complex.

**Reply by asturur** [reply]
This is  made thinking at the offset point relative to the rect center, but if the offset point is absolute on the canvas and external to the rect it works anyway, just some extra math is required

**Reply by theredsox** [reply]
Hi @asturur, thanks for your example. Much easier strategy, acting upon the object relatively. I like that it makes the current angle in the animation irrelevant. My offset/pivot point is always internal to the object so that should work as-is.

**Reply by asturur** [reply]
even if it was external would work, until is an offset it will work.
If you want to rotate around point 5, 5 on the canvas indipendently from where the object is ( for example ) , then you have to do some different work
